File: src/evidence/mock-inspection.ts
import { readFileSync, readdirSync } from "node:fs";
import { extname, join, relative } from "node:path";

import type { MockInjection, MockInspectionRecord } from "./types.js";
import { listRecordedSuites, readMockInspection } from "./store.js";

const SOURCE_EXTENSIONS = new Set([".rs", ".py", ".ts", ".tsx", ".js", ".jsx"]);
const EXCLUDED_DIRECTORIES = new Set([
  ".git",
  "dist",
  "node_modules",
  "spec",
  "target",
  "vendor",
]);

interface TestMarker {
  at: number;
  end: number;
  symbol: string;
}

/** Exact-commit inspection input for the pure auditor. */
export function mockInspectionInput(
  repo: string,
  commit: string | undefined,
): { injections: MockInjection[]; suites: string[] } {
  if (!commit) return { injections: [], suites: [] };
  const records = listRecordedSuites(repo)
    .map((suite) => readMockInspection(repo, suite, commit))
    .filter(
      (record): record is MockInspectionRecord =>
        record !== null && record.commit === commit,
    );
  return {
    suites: records.map((record) => record.suite).sort(compare),
    injections: records
      .flatMap((record) => record.injections)
      .sort(
        (a, b) =>
          compare(a.suite, b.suite) ||
          compare(a.path ?? "", b.path ?? "") ||
          (a.line ?? 0) - (b.line ?? 0),
      ),
  };
}

/**
 * Inspect test source for explicit stand-in constructors.
 *
 * This is deliberately narrow. It recognizes names declaring their intent
 * (`Mock*`, `Fake*`, `Stub*`) and explicit permissive factories such as
 * `Confirmation::allow()`. The auditor independently decides whether that
 * identifier overlaps the behaviour an obligation claims to verify.
 */
export function inspectMockInjections(
  repo: string,
  suite: string,
): MockInjection[] {
  const out: MockInjection[] = [];
  for (const path of sourceFiles(repo)) {
    const source = readFileSync(path, "utf8");
    const markers = testMarkers(source, extname(path));
    if (markers.length === 0) continue;

    const call =
      /\b([A-Z][A-Za-z0-9_]*(?:(?:::|\.)[A-Za-z_][A-Za-z0-9_]*)+)\s*\(/g;
    for (const match of source.matchAll(call)) {
      const identifier = match[1];
      if (!looksLikeStandIn(identifier)) continue;
      const marker = nearestMarker(markers, match.index ?? -1);
      if (!marker) continue;
      out.push({
        suite,
        symbol: marker.symbol,
        injects: [identifier],
        path: relative(repo, path),
        line: source.slice(0, match.index).split("\n").length,
      });
    }
  }

  const merged = new Map<string, MockInjection>();
  for (const item of out) {
    const key = `${item.path ?? ""}:${item.line ?? 0}:${item.symbol}`;
    const prior = merged.get(key);
    if (prior) {
      prior.injects = [...new Set([...prior.injects, ...item.injects])].sort();
    } else {
      merged.set(key, { ...item, injects: [...item.injects].sort() });
    }
  }
  return [...merged.values()].sort(
    (a, b) =>
      compare(a.path ?? "", b.path ?? "") ||
      (a.line ?? 0) - (b.line ?? 0) ||
      compare(a.symbol, b.symbol),
  );
}

function sourceFiles(root: string): string[] {
  const out: string[] = [];
  const walk = (dir: string): void => {
    for (const entry of readdirSync(dir, { withFileTypes: true })) {
      if (entry.isDirectory() && EXCLUDED_DIRECTORIES.has(entry.name)) continue;
      const path = join(dir, entry.name);
      if (entry.isDirectory()) walk(path);
      else if (entry.isFile() && SOURCE_EXTENSIONS.has(extname(path)))
        out.push(path);
    }
  };
  walk(root);
  return out.sort(compare);
}

function testMarkers(source: string, extension: string): TestMarker[] {
  const markers: TestMarker[] = [];
  const patterns =
    extension === ".rs"
      ? [
          /#\s*\[\s*test\s*\][\s\S]{0,240}?\bfn\s+([A-Za-z_][A-Za-z0-9_]*)\s*\(/g,
        ]
      : extension === ".py"
        ? [/(?:^|\n)\s*(?:async\s+)?def\s+(test_[A-Za-z0-9_]*)\s*\(/g]
        : [
            /\b(?:test|it)\s*\(\s*["'`]([^"'`]+)["'`]/g,
            /\b(?:async\s+)?function\s+(test[A-Za-z0-9_]*)\s*\(/g,
          ];
  for (const pattern of patterns) {
    for (const match of source.matchAll(pattern)) {
      const matchAt = match.index ?? 0;
      // The Python pattern consumes the preceding newline. Starting the span
      // there makes pythonFunctionEnd see that same newline as the end of the
      // signature and terminate before the function body. Anchor on the
      // captured function name instead.
      const at =
        extension === ".py"
          ? matchAt + match[0].lastIndexOf(match[1])
          : matchAt;
      const end =
        extension === ".py"
          ? pythonFunctionEnd(source, at)
          : bracedBodyEnd(source, at + match[0].length);
      markers.push({ at, end, symbol: match[1] });
    }
  }
  return markers.sort((a, b) => a.at - b.at);
}

function nearestMarker(markers: TestMarker[], at: number): TestMarker | null {
  let found: TestMarker | null = null;
  for (const marker of markers) {
    if (marker.at > at) break;
    if (at <= marker.end) found = marker;
  }
  return found;
}

function pythonFunctionEnd(source: string, at: number): number {
  const lineStart = source.lastIndexOf("\n", at) + 1;
  const indentation = /^\s*/.exec(source.slice(lineStart))?.[0].length ?? 0;
  let next = source.indexOf("\n", at);
  if (next < 0) return source.length;
  next += 1;
  while (next < source.length) {
    const end = source.indexOf("\n", next);
    const lineEnd = end < 0 ? source.length : end;
    const line = source.slice(next, lineEnd);
    if (line.trim() && !line.trimStart().startsWith("#")) {
      const current = /^\s*/.exec(line)?.[0].length ?? 0;
      if (current <= indentation) return next - 1;
    }
    next = lineEnd + 1;
  }
  return source.length;
}

/** Find the closing brace while ignoring braces inside strings and comments. */
function bracedBodyEnd(source: string, from: number): number {
  const open = source.indexOf("{", from);
  if (open < 0 || open - from > 500) return from;
  let depth = 0;
  let quote: string | null = null;
  let escaped = false;
  let lineComment = false;
  let blockComment = false;
  for (let index = open; index < source.length; index += 1) {
    const character = source[index];
    const next = source[index + 1] ?? "";
    if (lineComment) {
      if (character === "\n") lineComment = false;
      continue;
    }
    if (blockComment) {
      if (character === "*" && next === "/") {
        blockComment = false;
        index += 1;
      }
      continue;
    }
    if (quote) {
      if (escaped) escaped = false;
      else if (character === "\\") escaped = true;
      else if (character === quote) quote = null;
      continue;
    }
    if (character === "/" && next === "/") {
      lineComment = true;
      index += 1;
    } else if (character === "/" && next === "*") {
      blockComment = true;
      index += 1;
    } else if (character === '"' || character === "'" || character === "`") {
      quote = character;
    } else if (character === "{") {
      depth += 1;
    } else if (character === "}" && --depth === 0) {
      return index;
    }
  }
  return source.length;
}

function looksLikeStandIn(identifier: string): boolean {
  const parts = identifier.split(/::|\./);
  const type = parts[0];
  const method = parts.at(-1) ?? "";
  return (
    /^(?:mock|fake|stub|spy)/i.test(type) ||
    /^(?:allow|approve|approved|bypass|fake|mock|stub|succeed|success)$/i.test(
      method,
    )
  );
}

function compare(a: string, b: string): number {
  return a === b ? 0 : a < b ? -1 : 1;
}

File: rust/crates/quoin-evidence/src/mock_inspection.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Agent-IX

//! Source-level inspection for explicit stand-in substitutions (FR-032).
//!
//! Deliberately narrow. It recognizes names that declare their intent —
//! `Mock*`, `Fake*`, `Stub*`, `Spy*` — and explicit permissive factories such as
//! `Confirmation::allow()`. It does not decide whether that identifier overlaps
//! the behaviour an obligation claims to verify; the auditor does, from the
//! record this writes.
//!
//! # Offsets are bytes here and UTF-16 units in the oracle
//!
//! The retained TypeScript indexes with JavaScript string offsets. Every
//! comparison below is between offsets from the same scan, so the two agree
//! wherever the source is ASCII — which is every marker pattern and every
//! identifier the scan can match. The two bounded windows (`240` characters
//! between `#[test]` and its `fn`, `500` before an opening brace) are the one
//! place a multi-byte character shortens the Rust window, and both are
//! heuristics rather than contracts.

use std::collections::BTreeMap;

use regex::Regex;

use crate::error::EvidenceError;
use crate::ids::{Commit, SuiteId, SymbolId};
use crate::source::EvidenceSource;
use crate::store::{list_recorded_suites, read_mock_inspection};
use crate::types::MockInjection;

/// One test function's span in a source file.
#[derive(Debug, Clone)]
struct TestMarker {
    at: usize,
    end: usize,
    symbol: String,
}

/// Exact-commit inspection input for the pure auditor.
///
/// Only records **at that commit** count. An inspection of an older tree says
/// nothing about the source as it stands, and accepting one would let a stale
/// record answer for code it never read.
///
/// # Errors
///
/// [`EvidenceError::StoreIo`] or [`EvidenceError::StoreRead`] from the store.
pub fn mock_inspection_input<S: EvidenceSource + ?Sized>(
    source: &S,
    commit: Option<&Commit>,
) -> Result<(Vec<SuiteId>, Vec<MockInjection>), EvidenceError> {
    let Some(commit) = commit else {
        return Ok((Vec::new(), Vec::new()));
    };
    let mut suites = Vec::new();
    let mut injections = Vec::new();
    for suite in list_recorded_suites(source)? {
        let Some(record) = read_mock_inspection(source, &suite, commit)? else {
            continue;
        };
        if &record.commit != commit {
            continue;
        }
        suites.push(record.suite.clone());
        injections.extend(record.injections);
    }
    suites.sort();
    injections.sort_by(|a, b| {
        a.suite
            .as_str()
            .cmp(b.suite.as_str())
            .then_with(|| {
                a.path
                    .as_deref()
                    .unwrap_or("")
                    .cmp(b.path.as_deref().unwrap_or(""))
            })
            .then_with(|| a.line.unwrap_or(0).cmp(&b.line.unwrap_or(0)))
    });
    Ok((suites, injections))
}

/// Inspect the repository's test source for explicit stand-in constructors.
///
/// # Errors
///
/// [`EvidenceError::StoreIo`] when the walk or a read fails.
pub fn inspect_mock_injections<S: EvidenceSource + ?Sized>(
    source: &S,
    suite: &SuiteId,
) -> Result<Vec<MockInjection>, EvidenceError> {
    let call = call_regex();
    let mut found: Vec<MockInjection> = Vec::new();
    for path in source.source_files()? {
        let text = source.source_text(&path)?;
        let markers = test_markers(&text, extension_of(&path));
        if markers.is_empty() {
            continue;
        }
        for capture in call.captures_iter(&text) {
            let Some(identifier) = capture.get(1) else {
                continue;
            };
            if !looks_like_stand_in(identifier.as_str()) {
                continue;
            }
            let Some(whole) = capture.get(0) else {
                continue;
            };
            let Some(marker) = nearest_marker(&markers, whole.start()) else {
                continue;
            };
            found.push(MockInjection {
                suite: suite.clone(),
                symbol: SymbolId::new(marker.symbol.clone()),
                injects: vec![identifier.as_str().to_owned()],
                path: Some(path.clone()),
                line: Some(line_of(&text, whole.start())),
            });
        }
    }

    // Merged on `path:line:symbol`: one call site naming two stand-ins is one
    // injection carrying both, not two injections a reader has to reassemble.
    let mut merged: BTreeMap<String, MockInjection> = BTreeMap::new();
    for item in found {
        let key = format!(
            "{}:{}:{}",
            item.path.as_deref().unwrap_or(""),
            item.line.unwrap_or(0),
            item.symbol
        );
        merged
            .entry(key)
            .and_modify(|prior| {
                prior.injects.extend(item.injects.iter().cloned());
                prior.injects.sort();
                prior.injects.dedup();
            })
            .or_insert_with(|| {
                let mut fresh = item;
                fresh.injects.sort();
                fresh.injects.dedup();
                fresh
            });
    }
    let mut out: Vec<MockInjection> = merged.into_values().collect();
    out.sort_by(|a, b| {
        a.path
            .as_deref()
            .unwrap_or("")
            .cmp(b.path.as_deref().unwrap_or(""))
            .then_with(|| a.line.unwrap_or(0).cmp(&b.line.unwrap_or(0)))
            .then_with(|| a.symbol.as_str().cmp(b.symbol.as_str()))
    });
    Ok(out)
}

/// A qualified call: a capitalized type, one or more `::`/`.` segments, `(`.
fn call_regex() -> Regex {
    #[expect(
        clippy::expect_used,
        reason = "a literal pattern that fails to compile is a build-time defect, not a runtime condition, and the test below exercises it"
    )]
    Regex::new(r"\b([A-Z][A-Za-z0-9_]*(?:(?:::|\.)[A-Za-z_][A-Za-z0-9_]*)+)\s*\(")
        .expect("the call pattern is a literal")
}

fn extension_of(path: &str) -> &str {
    path.rsplit_once('.').map_or("", |(_, extension)| extension)
}

/// The 1-based line an offset falls on.
fn line_of(text: &str, at: usize) -> u64 {
    let prefix = text.get(..at).unwrap_or(text);
    u64::try_from(prefix.matches('\n').count() + 1).unwrap_or(u64::MAX)
}

/// Every test function in one file, ordered by position.
fn test_markers(text: &str, extension: &str) -> Vec<TestMarker> {
    #[expect(
        clippy::expect_used,
        reason = "literal patterns that fail to compile are a build-time defect, not a runtime condition"
    )]
    let patterns: Vec<Regex> = match extension {
        "rs" => vec![
            Regex::new(r"(?s)#\s*\[\s*test\s*\].{0,240}?\bfn\s+([A-Za-z_][A-Za-z0-9_]*)\s*\(")
                .expect("the rust test pattern is a literal"),
        ],
        "py" => vec![
            Regex::new(r"(?:^|\n)\s*(?:async\s+)?def\s+(test_[A-Za-z0-9_]*)\s*\(")
                .expect("the python test pattern is a literal"),
        ],
        _ => vec![
            Regex::new("\\b(?:test|it)\\s*\\(\\s*[\"'`]([^\"'`]+)[\"'`]")
                .expect("the javascript test pattern is a literal"),
            Regex::new(r"\b(?:async\s+)?function\s+(test[A-Za-z0-9_]*)\s*\(")
                .expect("the javascript function pattern is a literal"),
        ],
    };

    let mut markers = Vec::new();
    for pattern in &patterns {
        for capture in pattern.captures_iter(text) {
            let (Some(whole), Some(name)) = (capture.get(0), capture.get(1)) else {
                continue;
            };
            // The Python pattern consumes the preceding newline. Starting the
            // span there makes `python_function_end` see that same newline as
            // the end of the signature and terminate before the body, so the
            // span is anchored on the captured name instead.
            let at = if extension == "py" {
                name.start()
            } else {
                whole.start()
            };
            let end = if extension == "py" {
                python_function_end(text, at)
            } else {
                braced_body_end(text, whole.end())
            };
            markers.push(TestMarker {
                at,
                end,
                symbol: name.as_str().to_owned(),
            });
        }
    }
    markers.sort_by_key(|marker| marker.at);
    markers
}

/// The innermost test span containing an offset, or `None` outside every test.
fn nearest_marker(markers: &[TestMarker], at: usize) -> Option<&TestMarker> {
    let mut found = None;
    for marker in markers {
        if marker.at > at {
            break;
        }
        if at <= marker.end {
            found = Some(marker);
        }
    }
    found
}

/// Where a Python function body ends: the first later line indented no further.
fn python_function_end(text: &str, at: usize) -> usize {
    let line_start = text
        .get(..at)
        .and_then(|prefix| prefix.rfind('\n').map(|index| index + 1))
        .unwrap_or(0);
    let indentation = text
        .get(line_start..)
        .map_or(0, |rest| rest.len() - rest.trim_start().len());
    let Some(newline) = text.get(at..).and_then(|rest| rest.find('\n')) else {
        return text.len();
    };
    let mut next = at + newline + 1;
    while next < text.len() {
        let rest = text.get(next..).unwrap_or("");
        let line_end = rest.find('\n').map_or(text.len(), |index| next + index);
        let line = text.get(next..line_end).unwrap_or("");
        let trimmed = line.trim_start();
        if !trimmed.is_empty() && !trimmed.starts_with('#') {
            let current = line.len() - trimmed.len();
            if current <= indentation {
                return next.saturating_sub(1);
            }
        }
        next = line_end + 1;
    }
    text.len()
}

/// The closing brace of a braced body, ignoring braces in strings and comments.
fn braced_body_end(text: &str, from: usize) -> usize {
    let Some(relative) = text.get(from..).and_then(|rest| rest.find('{')) else {
        return from;
    };
    let open = from + relative;
    if open - from > 500 {
        return from;
    }
    let bytes = text.as_bytes();
    let mut depth = 0_i32;
    let mut quote: Option<u8> = None;
    let mut escaped = false;
    let mut line_comment = false;
    let mut block_comment = false;
    let mut index = open;
    while index < bytes.len() {
        let Some(&character) = bytes.get(index) else {
            break;
        };
        let next = bytes.get(index + 1).copied().unwrap_or(0);
        if line_comment {
            if character == b'\n' {
                line_comment = false;
            }
        } else if block_comment {
            if character == b'*' && next == b'/' {
                block_comment = false;
                index += 1;
            }
        } else if let Some(open_quote) = quote {
            if escaped {
                escaped = false;
            } else if character == b'\\' {
                escaped = true;
            } else if character == open_quote {
                quote = None;
            }
        } else if character == b'/' && next == b'/' {
            line_comment = true;
            index += 1;
        } else if character == b'/' && next == b'*' {
            block_comment = true;
            index += 1;
        } else if character == b'"' || character == b'\'' || character == b'`' {
            quote = Some(character);
        } else if character == b'{' {
            depth += 1;
        } else if character == b'}' {
            depth -= 1;
            if depth == 0 {
                return index;
            }
        }
        index += 1;
    }
    text.len()
}

/// Whether an identifier declares itself a stand-in.
///
/// Two independent rules: the *type* announces it (`MockClock`), or the *method*
/// is an explicit permissive factory (`Confirmation::allow`). Either is enough;
/// neither is a judgement about whether the substitution matters.
fn looks_like_stand_in(identifier: &str) -> bool {
    let parts: Vec<&str> = identifier
        .split("::")
        .flat_map(|part| part.split('.'))
        .collect();
    let type_name = parts.first().copied().unwrap_or("");
    let method = parts.last().copied().unwrap_or("");
    let lowered_type = type_name.to_ascii_lowercase();
    let lowered_method = method.to_ascii_lowercase();
    ["mock", "fake", "stub", "spy"]
        .iter()
        .any(|prefix| lowered_type.starts_with(prefix))
        || matches!(
            lowered_method.as_str(),
            "allow"
                | "approve"
                | "approved"
                | "bypass"
                | "fake"
                | "mock"
                | "stub"
                | "succeed"
                | "success"
        )
}
